### PR TITLE
fix(stage-ui): interrupt active chat responses

### PR DESCRIPTION
**Pull request by luoling8192**
## What changed

While a reply is active, an empty composer now shows one stop action in Web, mobile, and the Electron chat window. That action cancels both the active LLM request and TTS playback.

Typing text or attaching an image replaces stop with send. Sending during an active reply first interrupts the current LLM and TTS work, then submits the new message as the next turn.

The stop action stays hidden while that replacement is waiting for cross-window cancellation. Repeated submits are ignored until the replacement starts, so another action cannot race ahead of the new send. Visible partial assistant output is kept in history and marked as interrupted.

Electron follower windows now route cancellation to the leader-owned chat runtime, so the visible stop control reaches the real request.

Remote context-bridge turns also run through that leader boundary. Cancellation from a producer or receiver is broadcast to every context, clears each matching mirrored stream, and only stops the correlated producer runtime. The producer retires the old turn before cancellation begins, and interruption follows the response-owning session even if the user has switched to another conversation.

The shared Web chat controls now use one outline icon style. The desktop Stage toolbar uses one shared button primitive for its size, corner radius, colors, hover, pressed, and keyboard-focus states.

## Visual changes

| Before | After |
|---|---|
|  |  |
| Web composer controls | Web composer controls |

| Before | After |
|---|---|
|  |  |
| Desktop Stage toolbar | Desktop Stage toolbar |

## Validation

- `pnpm typecheck` (52/52 tasks passed)
- `pnpm lint` (passes with 649 existing repository warnings and 0 errors)
- Stage Layouts interruption tests: 10 passed
- Stage UI composer tests: 6 passed
- Core agent runtime tests: 32 passed, including preservation of interrupted partial output
- Stage UI chat and context-bridge contract tests: 49 passed, including follower-to-leader send/cancellation, producer-originated cancellation, multi-receiver cleanup, and cancellation-order correlation
- Electron InteractiveArea browser tests: 32 passed
- Local Stage Web visual verification in dark mode
  - composer tools: 32 x 32 px, 6 px radius
  - desktop toolbar: 36 x 36 px, 8 px radius

### CODE REVIEW

File: packages/stage-layouts/src/composables/use-chat-interruption.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+  const responseActive = computed(() => options.generating.value || showStopSpeakingButton.value)
+  const showStopAction = computed(() => responseActive.value && !options.hasSubmission.value)
+
+  async function cancelGeneration() {
+    await chatStore.cancelPendingSends(options.sessionId.value)
+  }
+
+  async function stopActiveResponse() {
+    stopSpeakingFromChat()
```
[the remark this answers is not in the document]
Comment (luoling8192) [reply packages/stage-layouts/src/composables/use-chat-interruption.ts:61]: Fixed in 008952ddf. Speech stop requests now use a same-origin BroadcastChannel, and a browser regression test verifies that a request from one renderer reaches the Stage output owner in another renderer.

File: packages/stage-ui/src/composables/use-data-maintenance.ts [hunk 2]
```diff
@@ -80,8 +80,8 @@ export function useDataMaintenance() {
       throw failure.reason
   }
 
-  function deleteAllChatSessions() {
-    chatOrchestrator.cancelPendingSends()
+  async function deleteAllChatSessions() {
+    await chatOrchestrator.cancelPendingSends()
```
[the remark this answers is not in the document]
Comment (luoling8192) [reply packages/stage-ui/src/composables/use-data-maintenance.ts:84]: Fixed in 008952ddf. `deleteAllChatSessions` now awaits both cancellation and `resetAllSessions`; `deleteAllData` and the chats settings action await the full operation before reporting success.

File: packages/stage-layouts/src/components/Layouts/MobileInteractiveArea.vue [hunk 3]
```diff
@@ -71,6 +72,15 @@ const {
 } = composer
 const { addFiles, selectFiles, error: imageError, pending: pendingImages } = useChatImages(composer, () => activeSessionId.value)
 const imageInput = useTemplateRef<HTMLInputElement>('imageInput')
+const hasSubmission = computed(() => !!messageInput.value.trim() || attachments.value.length > 0)
+const { showStopAction, stopActiveResponse, submitInterruptingResponse } = useChatInterruption({
+  sessionId: activeSessionId,
+  generating: isActiveSessionSending,
```
[the remark this answers is not in the document]
Comment (luoling8192) [reply packages/stage-layouts/src/components/Layouts/MobileInteractiveArea.vue:78]: Fixed in 008952ddf. Context-bridge streams now carry a correlated `{ sessionId, turnId }` cancellation command back to the renderer that produced the turn, where a local-only cancellation action aborts that renderer's runtime. Browser contract tests cover both command emission and producer-side cancellation.

File: packages/stage-layouts/src/composables/use-chat-interruption.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    await cancelGeneration()
+  }
+
+  async function submitInterruptingResponse() {
+    if (responseActive.value) {
+      interruptSpeakingFromChat()
+      await cancelGeneration()
+    }
+
+    await options.submit()
```
[the remark this answers is not in the document]
Comment (luoling8192) [reply packages/stage-layouts/src/composables/use-chat-interruption.ts]: Fixed in 008952ddf. The composer now captures its immutable submission and session before running the asynchronous interruption hook. A regression test switches sessions during the hook and verifies the original snapshot is sent to the original session.

File: packages/stage-layouts/src/composables/use-chat-interruption.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    await chatStore.cancelPendingSends(options.sessionId.value)
+  }
+
+  async function stopActiveResponse() {
+    stopSpeakingFromChat()
+    await cancelGeneration()
+  }
+
+  async function submitInterruptingResponse() {
+    if (responseActive.value) {
```
[the remark this answers is not in the document]
Comment (luoling8192) [reply packages/stage-layouts/src/composables/use-chat-interruption.ts]: Fixed in 008952ddf. Interruption now runs as the composer's `beforeSend` hook, which is invoked only after non-empty submission validation. Empty keyboard submits remain ignored and do not cancel generation; the interruption test covers this path.

File: packages/stage-ui/src/stores/mods/api/context-bridge.ts [hunk 6]
```diff
@@ -101,8 +101,20 @@ export const useContextBridgeStore = defineStore('mods:api:context-bridge', () =
...
 
+  async function cancelRemoteStream(sessionId: string) {
+    const guard = remoteStreamGuard.value
+    if (!guard || guard.sessionId !== sessionId)
+      return
+
+    await contextChannel?.emitStreamCancel({
+      sessionId: guard.sessionId,
+      turnId: guard.turnId,
+    })
```
[the remark this answers is not in the document]
Comment (luoling8192) [reply packages/stage-ui/src/stores/mods/api/context-bridge.ts:119]: Fixed in 26323f848. Cancelling a correlated mirrored turn now resets the visible foreground stream and clears the remote stream guard after the cancel command is published. The browser contract test verifies `isReceivingRemoteStream` returns to false and the stream is reset.

File: packages/stage-layouts/src/composables/use-chat-interruption.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  const {
+    interruptSpeakingFromChat,
+    showStopSpeakingButton,
+    stopSpeakingFromChat,
+  } = useStopSpeakingButton()
+
+  const responseActive = computed(() => options.generating.value || showStopSpeakingButton.value)
+  const showStopAction = computed(() => responseActive.value && !options.hasSubmission.value)
+
+  async function cancelGeneration(sessionId = options.sessionId.value) {
```
[the remark this answers is not in the document]
Comment (luoling8192) [reply packages/stage-layouts/src/composables/use-chat-interruption.ts]: Fixed in 26323f848. Interruption now resolves the response owner from the correlated remote stream first, then the active send session, and only falls back to the selected session. Regression tests cover stopping and sending an interruption after switching to another chat.

File: packages/stage-ui/src/stores/chat.ts [hunk 8]
```diff
@@ -577,7 +577,12 @@ export const useChatStore = defineStore('chat', () => {
...
   }
 
-  function cancelPendingSends(sessionId?: string) {
+  async function cancelPendingSends(sessionId?: string) {
+    runtime.cancelPendingSends(sessionId)
+  }
+
+  /** Cancels work owned by this renderer without leader RPC routing. */
+  function cancelLocalPendingSends(sessionId?: string) {
     runtime.cancelPendingSends(sessionId)
```
[the remark this answers is not in the document]
Comment (luoling8192) [reply packages/stage-ui/src/stores/chat.ts]: Fixed in 26323f848. Context-bridge input now uses the synchronized, serializable `send` action, so generation always runs in the leader-owned runtime. Correlated cancellation uses the synchronized `cancelPendingSends` action; the follower-to-leader contract test now starts and cancels the request from the follower.

File: packages/stage-layouts/src/composables/use-chat-interruption.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    await cancelGeneration()
+  }
+
+  async function interruptBeforeSend(sessionId: string) {
+    interruptSpeakingFromChat()
+    await cancelGeneration(sessionId)
+  }
+
+  async function submitInterruptingResponse() {
+    await options.submit(responseActive.value ? interruptBeforeSend : undefined)
```
[the remark this answers is not in the document]
Comment (luoling8192) [reply packages/stage-layouts/src/composables/use-chat-interruption.ts]: Fixed in 0e4d01f35. An interrupting submission now tracks its preparation and send-start boundaries. Stop stays hidden while cross-window cancellation is pending, then targets the replacement only after `send` has started. Regression tests cover the cancellation gap.

File: packages/stage-layouts/src/composables/use-chat-interruption.ts [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    stopSpeakingFromChat,
+  } = useStopSpeakingButton()
+
+  const responseActive = computed(() => options.generating.value || showStopSpeakingButton.value)
+  const showStopAction = computed(() => responseActive.value && !options.hasSubmission.value)
+
+  async function cancelGeneration(sessionId = options.sessionId.value) {
+    await Promise.all([
+      contextBridgeStore.cancelRemoteStream(sessionId),
+      chatStore.cancelPendingSends(sessionId),
```
[the remark this answers is not in the document]
Comment (luoling8192) [reply packages/stage-layouts/src/composables/use-chat-interruption.ts:55]: Fixed in 0e4d01f35. A user-aborted generation now commits any visible partial assistant output with `interrupted: true` before clearing the foreground stream. The runtime regression test verifies the partial response remains in session history.

File: packages/stage-ui/src/stores/mods/api/context-bridge.ts [hunk 11]
```diff
@@ -491,6 +512,13 @@ export const useContextBridgeStore = defineStore('mods:api:context-bridge', () =
       })
       disposeHookFns.value.push(stopContextUpdates)
 
+      disposeHookFns.value.push(contextChannel.onStreamCancel(async (command) => {
+        if (localProducedStream?.sessionId !== command.sessionId || localProducedStream.turnId !== command.turnId)
+          return
```
[the remark this answers is not in the document]
Comment (luoling8192) [reply packages/stage-ui/src/stores/mods/api/context-bridge.ts:531]: Fixed in 495fbc508. Every context receiving a matching stream-cancel command now clears its mirrored guard and visible stream, while only the correlated producer cancels the runtime. The browser contract test covers a non-producing receiver.

File: packages/stage-ui/src/stores/mods/api/context-bridge.ts [hunk 12]
```diff
@@ -491,6 +512,13 @@ export const useContextBridgeStore = defineStore('mods:api:context-bridge', () =
...
+
+        await chatOrchestrator.cancelPendingSends(command.sessionId)
```
[the remark this answers is not in the document]
Comment (luoling8192) [reply packages/stage-ui/src/stores/mods/api/context-bridge.ts:534]: Fixed in 495fbc508. The producer correlation is cleared synchronously before awaiting runtime cancellation, so a duplicate or delayed command cannot reuse the old turn ownership while cancellation settles. The browser contract test holds cancellation pending and verifies the old command is handled once.

File: packages/stage-ui/src/stores/mods/api/context-bridge.ts [hunk 13]
```diff
@@ -100,9 +99,31 @@ export const useContextBridgeStore = defineStore('mods:api:context-bridge', () =
...
   const isReceivingRemoteStream = computed(() => remoteStreamGuard.value?.sessionId === chatSession.activeSessionId)
+  const remoteStreamSessionId = computed(() => remoteStreamGuard.value?.sessionId)
   let contextChannel: ReturnType<typeof createContextChannel> | undefined
+  let localProducedStream: { sessionId: string, turnId: string } | undefined
   let initialized = false
 
+  async function cancelRemoteStream(sessionId: string) {
+    const guard = remoteStreamGuard.value
+    if (!guard || guard.sessionId !== sessionId)
+      return
```
[the remark this answers is not in the document]
Comment (luoling8192) [reply packages/stage-ui/src/stores/mods/api/context-bridge.ts:112]: Fixed in 5b20f8517. `cancelRemoteStream` now broadcasts the correlated local producer turn as well as mirrored turns before the caller cancels the runtime. The browser contract test verifies producer-originated Stop reaches peer receivers.

File: packages/stage-layouts/src/composables/use-chat-interruption.ts [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+    await cancelGeneration(interruptedSessionId)
+  }
+
+  async function submitInterruptingResponse() {
+    if (!responseActive.value) {
+      await options.submit()
+      return
+    }
+
+    preparingReplacement.value = true
```
[the remark this answers is not in the document]
Comment (luoling8192) [reply packages/stage-layouts/src/composables/use-chat-interruption.ts:81]: Fixed in 5b20f8517. Replacement submission now rejects reentrant calls while its cancellation boundary is pending, so a duplicate Enter cannot clear the first call's stop suppression. The interruption regression test holds cancellation pending and verifies only one submission starts.

**Comment by dasilva333** [comment]
Nice work landing this in #2624 and #2623!

We recently designed and shipped this exact in-flight stop / cancellation pattern in our fork last month, and arrived at nearly the identical architecture (preserving partial streaming output with an interrupted flag, instant TTS intent cancellation, and morphing the send button).

In case it's helpful as a reference for follow-ups or edge-case audits, our full design journal and pitfall catalog is documented here:
🔗 https://github.com/dasilva333/airi/blob/main/docs/design-chat-stop-generation.md

A couple of subtle edge cases from our implementation that might be worth keeping in mind as people test:
1. **Outbound hook suppression**: Ensuring turn-complete hooks (`chat-turn-complete`) do not fire on abort, so truncated partial turns aren't relayed to outbound channels (like Discord bots) or ingested into long-term memory journals.
2. **Mid-tool loop aborts**: Handling stop events when an LLM is mid-flight between tool call markers / tool execution without hanging the `sending` state.
3. **Catch guard against fake error bubbles**: Making sure the catch block recognizes intentional aborts so it doesn't pop an error bubble or restore the sent draft back into the composer.

Kudos on getting this into upstream!
